server/src/main/java/io/crate/planner/operators/LogicalPlanner.java

```diff
@@ -109,13 +109,15 @@
 import io.crate.planner.optimizer.rule.RewriteNestedLoopJoinToHashJoin;
 import io.crate.planner.optimizer.rule.RewriteToQueryThenFetch;
 import io.crate.planner.optimizer.rule.SwapHashJoin;
+import io.crate.planner.optimizer.rule.SwapNestedLoopJoin;
 import io.crate.types.DataTypes;
 
 /**
  * Planner which can create a {@link ExecutionPlan} using intermediate {@link LogicalPlan} nodes.
  */
 public class LogicalPlanner {
     private final IterativeOptimizer optimizer;
+    private final IterativeOptimizer joinOptimizer;
```

Comment:
The idea of the join optimizer is that we do join re-ordering as a second step once  the logical plans are final. This is a lot easier because it results in much less dependencies between rules. Therefore two optimizers.

Reply:
Yes, the reordering would not cause other rules to get reapplied again. But the plans have to final to be able make the re-ordering. I can move that into it's own pr merging this before that with a clear commit message.